README.md

```diff
@@ -28,14 +28,245 @@ In addition, that format would include optional **signing** of the resources, wh
+the [Web App Manifest](https://www.w3.org/TR/appmanifest/), or does the fact
+that we're going to sign this rule out JSON?
+
+Resources are described by a line consisting of:
+
+```
+relative/url hashalgorithm-base64digest hashalgorithm2-base64digest2 ...
+```
+
+The URL is relative to the manifest's `Content-Location`.
+
+The hashes are in the format defined for the
+[`integrity` HTML attribute](https://w3c.github.io/webappsec-subresource-integrity/#the-integrity-attribute).
+This allows multiple hashes in order to provide agility in the face of future
+cryptographic discoveries. See also https://tools.ietf.org/html/rfc7696.
+
+The hashes are computed over the entire resource, including its length, HTTP
+headers, and body of the named resource. This differs
+from [SRI](https://w3c.github.io/webappsec-subresource-integrity), which only
+hashes the body. Note: This will usually prevent a package from relying on some
+of its contents being transferred as normal network responses, unless its author
+can guarantee the network won't change or reorder the headers.
+
+TODO: Do we need the length to be shown in the manifest, or is its presence at
+the beginning of the resource, included in the hash, enough?
+
+
+#### Sub-packages
+
+There are three possible forms of dependencies on sub-packages, of which we
+allow two. Note that a sub-package is or can be protected by its
+own [signature](#signatures), so if the main package trusts the sub-package's
+server, it could avoid specifying a version of the sub-package at all. However,
+this opens the main package up to downgrade attacks, where the sub-package is
+replaced by an older, vulnerable version, so we don't allow this option.
+
+If the main package wants to load either the sub-package it was built with or
+any upgrade, it can specify the date of the original sub-package:
+
+```
+https://package/manifest/url Date: Tue, 07 Feb 2017 01:05:54 GMT
+```
+
+Constraining packages with their date makes it possible to link together
+sub-packages with common dependencies, even if the sub-packages were built at
+different times.
+
+If the main package wants to be certain it's loading the exact version of a
+sub-package that it was built with, it can constrain sub-package with a hash of its manifest:
+
+```
+https://package/manifest/url hashalgorithm-base64digest hashalgorithm2-base64digest2 ...
+```
+
+Note that because the sub-package may include sub-sub-packages by date, the top
+package may need to explicitly list those sub-sub-packages' hashes in order to
+be completely constrained.
+
+It should be possible to additionally constrain packages using semver, for the
+benefit of NPM and other packaging systems, but I haven't explored that
+extension here.
+
+
+#### Signatures
+
+A [manifest](#manifest) is signed by one or more `Signature` headers. The
+message to sign consists of the length, followed by the concatenation of all
+headers except any `Signature` headers, followed by the body of the manifest.
```

Comment:
This needs to be reworded, `The message to sign` is confusing, are we talking about the whole package, the manifest, or an individual resource. I assume we are only talking about the manifest.

The wording here makes it sound like after generating a manifest we put the signatures on the resources.